module04/01-Cash-Desk/solution.py

```diff
@@ -0,0 +1,98 @@
+class Bill:
+    def __init__(self, amount):
+        if not isinstance(amount, int):
+            raise TypeError('Amount is not integer')
+
+        if amount < 0:
+            raise ValueError('Amount is negative')
+        self.amount = amount
+
+    def __str__(self):
+        return "A " + str(self.amount) + "$ bill"
+
+    def __repr__(self):
+        return self.amount
+
+    def __int__(self):
+        return self.amount
+
+    def __eq__(self, other):
+        if self.__int__() == other.__int__():
+            return True
+        else:
+            return False
+
+    def __hash__(self):
+        return hash(self.amount)
+
+
+class BatchBill:
+    def __init__(self, bills):
+        self.bills = bills
+
+    def __int__(self):
+        return self.total()
+
+    def __len__(self):
+        return self.bills.count
+
+    def total(self):
+        s = 0
+        for bill in self.bills:
+            s += bill.amount
+        return s
+
+    def __getitem__(self, index):
+        return self.bills[index]
+
+
+class CashDesk:
+    def __init__(self):
+        self.banknotes = {}
+
+    def take_bill(self, bill):
+        amount = bill.amount
+        if not (amount in self.banknotes):
+            self.banknotes[amount] = 1
+        else:
+            self.banknotes[amount] += 1
+        return
+
+    def take_money(self, money):
+        if isinstance(money, Bill):
+            self.take_bill(money)
+        else:
+            for m in money:
+                self.take_bill(m)
+        return
+
+    def total(self):
+        s = 0
+        for banknote in self.banknotes:
+            count = self.banknotes[banknote];
+            s += count * banknote
+        return s
+
+    def inspect(self):
+        s = 'We have a total of ' + str(self.total()) + '$ in the desk' + '\n'
+        s += 'We have the following count of bills, sorted in ascending order:'
+        keys = sorted(self.banknotes.keys())
+        for k in keys:
+            s += '\n' + str(k) + '$ bills - ' + str(self.banknotes[k])
+        return s
+
+
+
```

Comment:
Please use `__main__` block for the extra code that you use to test for yourself while developing

Reply:
Good point!